### PR TITLE
Feature: Support configuring Kubernetes secret type

### PR DESCRIPTION
**Pull request by Marton6**
Based off of: https://github.com/1Password/onepassword-operator/pull/80/files

Adds support for configuring kubernetes secret type.

### CODE REVIEW

**Comment by Marton6** [comment]
Using secret types defined by kubernetes is also possible, but for this to work you need to make the structure of your 1Password item reflect the structure kubernetes expects.

File: pkg/kubernetessecrets/kubernetes_secrets_builder.go
Comment (SimonBarendse) [inline pkg/kubernetessecrets/kubernetes_secrets_builder.go]: Shouldn't we do a create + delete in this case?

File: pkg/kubernetessecrets/kubernetes_secrets_builder.go
Comment (SimonBarendse) [inline pkg/kubernetessecrets/kubernetes_secrets_builder.go]: What happens when secretType is not a known secret type string by Kubernetes? Do we need validation or does Kubernetes do this in a later step?

**Comment by SimonBarendse** [comment]
Awesome, thank you @amontalban! :heart:

---

Thank you @Marton6  for making some additional changes. I have a couple questions for these.

>  for this to work you need to make the structure of your 1Password item reflect the structure kubernetes expects.

How do we error when the structure does not match expectations? Is it clear from the message how to resolve this?

File: pkg/kubernetessecrets/kubernetes_secrets_builder.go
Comment (Marton6) [reply pkg/kubernetessecrets/kubernetes_secrets_builder.go]: According to the docs any string can be a valid secret type. The ones defined in the kubernetes docs are handled differently by kubernetes, but any user can define their own types and handle them differently from other secret types.

[Quote](https://kubernetes.io/docs/concepts/configuration/secret/#secret-types):
> The type of a Secret is used to facilitate programmatic handling of different kinds of confidential data.
> [...]
> You can define and use your own Secret type by assigning a non-empty string as the type value for a Secret object. An empty string is treated as an Opaque type. Kubernetes doesn't impose any constraints on the type name. However, if you are using one of the builtin types, you must meet all the requirements defined for that type.

File: pkg/kubernetessecrets/kubernetes_secrets_builder.go
Comment (SimonBarendse) [reply pkg/kubernetessecrets/kubernetes_secrets_builder.go]: Ahh right, if this is expected behavior in K8s, let's definitely stick with that :+1: 

I think I confused with Terraform behavior for a bit.

**Review by SimonBarendse — APPROVED** [review]
Good to go when you're happy with the answer you find to:

> How do we error when the structure does not match expectations? Is it clear from the message how to resolve this?

:rocket:

**Comment by Marton6** [comment]
> > for this to work you need to make the structure of your 1Password item reflect the structure kubernetes expects.
> 
> How do we error when the structure does not match expectations? Is it clear from the message how to resolve this?

Unfortunately the error here is not super clear and figuring out how to properly return errors to the user will take more time. I don't want to hold up this MR just for that, as you can make sense of what's wrong if you read the logs.

If the structure does not match what kubernetes expected, the following error is logged by the kubernetes operator:
```
{"level":"error","ts":1645184943.8607564,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"onepassworditem-controller","request":"default/tls-example-item","error":"Secret \"tls-example-item\" is invalid: data[tls.crt]: Required value","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/workspace/vendor/github.com/go-logr/zapr/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:258\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90"}
```

Specifically the `error` field is the one with the error message here: `"Secret \"tls-example-item\" is invalid: data[tls.crt]: Required value"`. In this case I removed the `tls.crt` field which was expected by kubernetes.

**Comment by Marton6** [comment]
> There are no conditions for the opaque secret that we supported before this MR, right? In other words: the less-than-ideal error message is only present in additions, we're not making any existing situations worse, right?

Yes, that is correct.

**Pull request merged**
